### ISSUE TITLE
ValueError: w2 should be positive

### ISSUE THREAD

**Issue by jcohenadad** [spinalcordtoolbox/spinalcordtoolbox#3703]
### Description

nibabel's `quaternions.py` function has a check to ensure the quaternions are correct wrt. the affine matrix. However, due to precision issues, this check could fail, hence the presence of the method `nib.Nifti1Header.quaternion_threshold`. In rare instances, the threshold is not sufficient and loading the data fails. 

Full context: https://forum.spinalcordmri.org/t/sct-crop-image-valueerror-w2-should-be-positive/826/10

Related issue: https://github.com/nipy/nibabel/issues/626

### Steps to Reproduce

This bug is data-dependent. The data were sent privately to @jcohenadad. 

### Proposed solutions

1. Add a user-accessible flag to control `nib.Nifti1Header.quaternion_threshold`. Cons: the flag would need to be added to all CLI functions, which is probably not a good idea.
2. Add an environment variable that would be read by SCT's `image.py` API. Cons: environment variables are not common in SCT (ie: users are not aware of them, and don't use them often).
3. Hard-code the line `nib.Nifti1Header.quaternion_threshold = -1e-06` when opening an image. Cons: (i) it bypasses nibabel's decision to _not_ set that threshold value; (ii) it _might_ have an impact on cross-compatibility.

I would personally go with (3).

**Comment by mguaypaq** [comment]
For context, here are the offending lines used by the image cropping operation:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/f1766ed3f63b4d83fecb7a382b357a2427d8e6c2/spinalcordtoolbox/cropping.py#L97-L104

This code updates the origin of the image after cropping, both in the q-form and the s-form. The problem arises when the call to `img_out.hdr.get_qform()` raises an exception, because the original image file's q-form isn't quite valid: it contains a quaternion whose norm is very slightly bigger than 1.

- By default, nibabel uses a cutoff of `img_out.quaternion_threshold = -np.finfo(np.float32).eps * 3` which is approx `-3.6e-07`, for the most negative value that gets automatically clamped to zero, when computing the (squared) real part of a quaternion from its imaginary parts, in `get_qform()`. It's not clear to me that the 3 in `-np.finfo(np.float32).eps * 3` is sufficient to cover the imprecision in the computation `1.0 - x*x - y*y - z*z`; there are 3 subtractions, but also squarings. I think an error of ε for each of x, y, z could get magnified to 2 * sqrt(3) * ε ~= 3.464 * ε.
- In the forum post, the computed squared real part of the quaternion is `w2 = -4.664622e-07`, which is *slightly* too negative.
- Proposed solution 3 would be to set the threshold slightly larger, at `-1e-6`. This would be approx 3 times bigger than the default threshold, so, a small multiple, not orders of magnitude more. The extra imprecision would only affect the translation part of the q-form.

In other words, I think the proposed solution 3 is reasonable.

I will note, however, that updating the s-form by using the q-form seems like a bug, since the NIfTI standard says that [the q-form and the s-form can represent different coordinate systems](https://nifti.nimh.nih.gov/nifti-1/documentation/faq#Q19). Thankfully, the s-form doesn't use quaternions, so it shouldn't run into this kind of arithmetic error.

**Comment by jcohenadad** [comment]
@mguaypaq thank you for the thorough digging! Here is the full header (incl. the qform) of the 'problematic' image:

~~~
julien-macbook:~/Desktop/20220223_sct $ fslhd im.nii.gz 
filename	im.nii.gz
size of header	348
data_type	INT16
dim0		4
dim1		112
dim2		112
dim3		24
dim4		400
dim5		1
dim6		1
dim7		1
vox_units	mm
time_units	s
datatype	4
nbyper		2
bitpix		16
pixdim0		0.000000
pixdim1		0.750000
pixdim2		0.750000
pixdim3		3.000000
pixdim4		1.500000
pixdim5		0.000000
pixdim6		0.000000
pixdim7		0.000000
vox_offset	352
cal_max		0.000000
cal_min		0.000000
scl_slope	1.000000
scl_inter	0.000000
phase_dim	2
freq_dim	1
slice_dim	3
slice_name	sequential_increasing
slice_code	1
slice_start	0
slice_end	0
slice_duration	0.000000
toffset		0.000000
intent		Unknown
intent_code	27552
intent_name	
intent_p1	0.000000
intent_p2	0.000000
intent_p3	0.000000
qform_name	Scanner Anat
qform_code	1
qto_xyz:1	-0.750000 0.000000 0.000000 43.441898 
qto_xyz:2	0.000000 0.731651 -0.659540 -43.354099 
qto_xyz:3	-0.000000 -0.164885 -2.926603 -48.886902 
qto_xyz:4	0.000000 0.000000 0.000000 1.000000 
qform_xorient	Right-to-Left
qform_yorient	Posterior-to-Anterior
qform_zorient	Superior-to-Inferior
sform_name	Scanner Anat
sform_code	1
sto_xyz:1	-0.750000 0.000000 -0.000000 43.441898 
sto_xyz:2	0.000000 0.731651 0.659539 -43.354099 
sto_xyz:3	0.000000 -0.164885 2.926600 -48.886902 
sto_xyz:4	0.000000 0.000000 0.000000 1.000000 
sform_xorient	Right-to-Left
sform_yorient	Posterior-to-Anterior
sform_zorient	Inferior-to-Superior
file_type	NIFTI-1+
file_code	1
descrip		TE=21;Time=115547.555;phase=1
aux_file	
~~~

**Comment by jcohenadad** [comment]
> I will note, however, that updating the s-form by using the q-form seems like a bug, since the NIfTI standard says that [the q-form and the s-form can represent different coordinate systems](https://nifti.nimh.nih.gov/nifti-1/documentation/faq#Q19).

I would not call it a bug but rather a decision made in order to deal with the negative impacts of having a different s/qform when 3rd party software packages that _only_ use the sform are used. Related discussions:
- https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3283
- https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3174
- https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3005
- https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3284
- https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3254
- https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3251
- https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3249
- https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3097
- https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/2106

Bottomline: the q/sform usage is a never-ending saga, and our philosophy is to navigate the water by taking decisions that will the least negatively affect our very specific user base (ie knowing what software packages the use, how they use it, etc.).

**Comment by mguaypaq** [comment]
Here is a hand-crafted artificial example which triggers the error unless `quaternion_threshold` is increased to `-1e-6`. I'll turn it into a unit test.

[bad_quaternion.nii.gz](https://github.com/spinalcordtoolbox/spinalcordtoolbox/files/8492834/bad_quaternion.nii.gz)
 
Example usage:
```
Python 3.7.11 (default, Jul 27 2021, 14:32:16) 
[GCC 7.5.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import nibabel as nib
>>> im = nib.load('bad_quaternion.nii.gz')
>>> im.get_qform()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mguaypaq/neuropoly/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/nibabel/nifti1.py", line 1831, in get_qform
    return self._header.get_qform(coded)
  File "/home/mguaypaq/neuropoly/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/nibabel/nifti1.py", line 915, in get_qform
    quat = self.get_qform_quaternion()
  File "/home/mguaypaq/neuropoly/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/nibabel/nifti1.py", line 890, in get_qform_quaternion
    return fillpositive(bcd, self.quaternion_threshold)
  File "/home/mguaypaq/neuropoly/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/nibabel/quaternions.py", line 99, in fillpositive
    raise ValueError(f'w2 should be positive, but is {w2:e}')
ValueError: w2 should be positive, but is -9.536743e-07
>>> im.header.quaternion_threshold = -1e-6
>>> im.get_qform()
array([[ 7.49998569e-01,  0.00000000e+00,  5.85936941e-03,
         1.00000000e+01],
       [ 0.00000000e+00, -7.50000000e-01,  0.00000000e+00,
         2.00000000e+01],
       [ 1.46484235e-03,  0.00000000e+00, -2.99999428e+00,
         3.00000000e+01],
       [ 0.00000000e+00,  0.00000000e+00,  0.00000000e+00,
         1.00000000e+00]])
>>> 
```

**Comment by joshuacwnewton** [comment]
Reopening this image because the user has [reported this same error](https://forum.spinalcordmri.org/t/sct-crop-image-valueerror-w2-should-be-positive/826/14) once more:

> Thanks for addressing this! And sorry for the long delay. I no longer get the error in sct_crop_image, but now (on v5.7) I do get the error in sct_fmri_moco.
> 
> ```
> Copy file_target to a temporary file...
> 
> Register. Loop across Z (note: there is only one Z if orientation is axial)
> Z=0/0:   0%|                                          | 0/167 [00:00<?, ?iter/s]/usr/local/sct_v5.7.0/bin/isct_antsApplyTransforms -d 3 -i /private/var/folders/9d/lt6rxcbx1dq2yxb70xf1x00c0000gn/T/sct-20221108120531.607714-moco-ahn7lc5g/data_T0000.nii -o /private/var/folders/9d/lt6rxcbx1dq2yxb70xf1x00c0000gn/T/sct-20221108120531.607714-moco-ahn7lc5g/data_T0000_moco.nii -t mat_final/mat.Z0000T0000Warp.nii.gz -r datasub_0_mean.nii.gz -n Linear # in /private/var/folders/9d/lt6rxcbx1dq2yxb70xf1x00c0000gn/T/sct-20221108120531.607714-moco-ahn7lc5g
> Z=0/0:   0%|                                          | 0/167 [00:01<?, ?iter/s]
> Traceback (most recent call last):
>   File "/usr/local/sct_v5.7.0/spinalcordtoolbox/scripts/sct_fmri_moco.py", line 201, in <module>
>     main(sys.argv[1:])
>   File "/usr/local/sct_v5.7.0/spinalcordtoolbox/scripts/sct_fmri_moco.py", line 186, in main
>     fname_output_image = moco_wrapper(param)
>   File "/usr/local/sct_v5.7.0/spinalcordtoolbox/moco.py", line 373, in moco_wrapper
>     file_mat_data, im_moco = moco(param_moco)
>   File "/usr/local/sct_v5.7.0/spinalcordtoolbox/moco.py", line 604, in moco
>     file_data_splitZ_splitT_moco[it], im_mask=input_mask)
>   File "/usr/local/sct_v5.7.0/spinalcordtoolbox/moco.py", line 754, in register
>     '-v', '0'])
>   File "/usr/local/sct_v5.7.0/spinalcordtoolbox/scripts/sct_apply_transfo.py", line 360, in main
>     transform.apply()
>   File "/usr/local/sct_v5.7.0/spinalcordtoolbox/scripts/sct_apply_transfo.py", line 294, in apply
>     im_src_reg.copy_qform_from_ref(Image(fname_dest))
>   File "/usr/local/sct_v5.7.0/spinalcordtoolbox/image.py", line 378, in copy_qform_from_ref
>     self.hdr.set_qform(im_ref.hdr.get_qform())
>   File "/usr/local/sct_v5.7.0/python/envs/venv_sct/lib/python3.7/site-packages/nibabel/nifti1.py", line 917, in get_qform
>     quat = self.get_qform_quaternion()
>   File "/usr/local/sct_v5.7.0/python/envs/venv_sct/lib/python3.7/site-packages/nibabel/nifti1.py", line 892, in get_qform_quaternion
>     return fillpositive(bcd, self.quaternion_threshold)
>   File "/usr/local/sct_v5.7.0/python/envs/venv_sct/lib/python3.7/site-packages/nibabel/quaternions.py", line 99, in fillpositive
>     raise ValueError(f'w2 should be positive, but is {w2:e}')
> ValueError: w2 should be positive, but is -4.147325e-07
> ```

There are 7 total places we're calling `.get_qform()`, and so I have some concerns about how feasible it would be to repeat the fix from #3771 in every location.

So, we could possibly set `quaternion_threshold` directly inside the `Image` when creating an `Image` object, rather than at the point of the `.get_qform` call? (This is assuming that we want this threshold to modified globally for all usages of `Image` objects, and that the header property of an `Image` object isn't ever replaced in its entirety.)
